packages/lib/gpc/src/gpc.ts

```diff
@@ -179,3 +183,77 @@ export async function gpcVerify(
     circuitOutputs
   );
 }
+
+/**
+ * Possible sources to download GPC artifacts.
+ */
+export type GPCArtifactSource = "zupass" | "github" | "unpkg";
+
+/**
+ * Stability level of GPC artifacts to use.  Test artifacts are for use
+ * in active development while prod artifacts are officially released.
+ */
+export type GPCArtifactStability = "prod" | "test";
+
+/**
+ * Version specifier for GPC artifacts.  It meaning depends on the source.
+ * It might be the version of an NPM package release (e.g. 1.0.1) or a GitHub
+ * revision identifier (branch, tag, or commit).
+ */
+export type GPCArtifactVersion = string | undefined;
+
+/**
+ * Forms a URL for downloading GPC artifacts depending on configuration
+ *
+ * @param source the download source location
+ * @param stability the stability level (test or prod) of artifacts to seek.
+ *   Ignored in some sources in favor of the version.
+ * @param version the version identifier for circuit artifacts.  Not relevant
+ *   to some sources which host only a single version.
+ * @param zupassURL the base URL for Zupass, if used as a download option.
+ *   Can be "" or "/" to use a relative URL (within the Zupass app).
+ * @returns a root URL to download GPC artifacts, as needed for {@link gpcProve}
+ *   or {@link gpcVerify}.
+ */
+export function gpcArtifactDownloadURL(
+  source: GPCArtifactSource,
+  stability: GPCArtifactStability,
+  version: GPCArtifactVersion,
+  zupassURL?: string
+): string {
+  switch (source) {
+    case "github":
+      const REPO_NAME = "proofcarryingdata/snark-artifacts";
+      if (version === undefined || version === "") {
+        throw new Error("GitHub artifact download requires a version.");
+      }
+      return githubDownloadRootURL(
+        REPO_NAME,
+        PROTO_POD_GPC_FAMILY_NAME,
+        version
+      );
+    case "unpkg":
+      if (version === undefined || version === "") {
+        throw new Error("unpkg artifact download requires a version.");
+      }
+      // stability is intentionally ignored.  NPM version can encode
+      // pre-release status.
+      return unpkgDownloadRootURL(PROTO_POD_GPC_FAMILY_NAME, version);
+    case "zupass":
+      // TODO(POD-P3): Do we want to expose source=zupass as a public option?
+      // If so, we need the Zupass server to not set `Access-Control-Allow-Origin: *`,
+      // or migrate to a different hosting option.
+      if (zupassURL === undefined) {
+        throw new Error(
+          'Zupass artifact download requires a server URL.  Try "https://zupass.org".'
+        );
+      }
+      return urljoin(
+        zupassURL,
+        stability === "test" ? "artifacts/test" : "artifacts",
+        PROTO_POD_GPC_FAMILY_NAME
+      );
+    default:
+      throw new Error(`Unknown artifact download source ${source}.`);
```

Comment:
in zupass client - if this throws - is it handled gracefully?

Reply:
It's a crash on page load, so not very graceful.  But it's based on bad deployment config on the server side, so I'm not sure how much more graceful it needs to be?  If someone's app is configuring this from client-side data, it's their problem to decide how to handle errors.